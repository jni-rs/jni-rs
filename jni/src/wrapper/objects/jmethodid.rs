use std::marker::PhantomData;

use sys::jmethodID;

/// Wrapper around `sys::jmethodid` that adds a lifetime. This prevents it from
/// outliving the context in which it was acquired and getting GC'd out from
/// under us. It matches C's representation of the raw pointer, so it can be
/// used in any of the extern function argument positions that would take a
/// `jmethodid`.
#[repr(C)]
#[derive(Copy, Clone, Debug)]
pub struct JMethodID<'a> {
    internal: jmethodID,
    lifetime: PhantomData<&'a ()>,
}

impl<'a> From<jmethodID> for JMethodID<'a> {
    fn from(other: jmethodID) -> Self {
        JMethodID {
            internal: other,
            lifetime: PhantomData,
        }
    }
}

impl<'a> JMethodID<'a> {
    /// Unwrap to the internal jni type.
    pub fn into_inner(self) -> jmethodID {
        self.internal
    }
}
