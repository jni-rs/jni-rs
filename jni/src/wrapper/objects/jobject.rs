use std::marker::PhantomData;

use sys::jobject;

/// Wrapper around `sys::jobject` that adds a lifetime. This prevents it from
/// outliving the context in which it was acquired and getting GC'd out from
/// under us. It matches C's representation of the raw pointer, so it can be
/// used in any of the extern function argument positions that would take a
/// `jobject`.
///
/// Most other types in the `objects` module deref to this, as they do in the C
/// representation.
#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct JObject<'a> {
    internal: jobject,
    lifetime: PhantomData<&'a ()>,
}

impl<'a> From<jobject> for JObject<'a> {
    fn from(other: jobject) -> Self {
        JObject {
            internal: other,
            lifetime: PhantomData,
        }
    }
}

impl<'a> ::std::ops::Deref for JObject<'a> {
    type Target = jobject;

    fn deref(&self) -> &Self::Target {
        &self.internal
    }
}

impl<'a> JObject<'a> {
    /// Unwrap to the internal jni type.
    pub fn into_inner(self) -> jobject {
        self.internal
    }

    /// Creates a new null object
    pub fn null() -> JObject<'a> {
        (::std::ptr::null_mut() as jobject).into()
    }
}
