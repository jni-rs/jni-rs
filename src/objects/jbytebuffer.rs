use std::{borrow::Cow, ops::Deref};

use crate::{
    objects::{Global, JClass, LoaderContext},
    Env,
};

struct JByteBufferAPI {
    class: Global<JClass<'static>>,
}

crate::define_reference_type!(
    JByteBuffer,
    "java.nio.ByteBuffer",
    |env: &mut Env, loader_context: &LoaderContext| {
        let class = loader_context.load_class_for_type::<JByteBuffer>(false, env)?;
        let class = env.new_global_ref(&class).unwrap();
        Ok(Self { class })
    }
);
