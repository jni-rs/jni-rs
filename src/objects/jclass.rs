use std::ops::Deref;

use once_cell::sync::OnceCell;

use crate::{
    env::Env,
    errors::Result,
    objects::{Global, JClassLoader, JMethodID, JObject, JStaticMethodID, JValue, LoaderContext},
    signature::JavaType,
    strings::JNIStr,
    sys::{jclass, jobject},
};

use super::Reference;

/// Lifetime'd representation of a `jclass`. Just a `JObject` wrapped in a new
/// class.
#[repr(transparent)]
#[derive(Debug, Default)]
pub struct JClass<'local>(JObject<'local>);

impl<'local> AsRef<JClass<'local>> for JClass<'local> {
    fn as_ref(&self) -> &JClass<'local> {
        self
    }
}

impl<'local> AsRef<JObject<'local>> for JClass<'local> {
    fn as_ref(&self) -> &JObject<'local> {
        self
    }
}

impl<'local> ::std::ops::Deref for JClass<'local> {
    type Target = JObject<'local>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<'local> From<JClass<'local>> for JObject<'local> {
    fn from(other: JClass) -> JObject {
        other.0
    }
}
struct JClassAPI {
    class: Global<JClass<'static>>,
    get_class_loader_method: JMethodID,
    for_name_method: JStaticMethodID,
    for_name_with_loader_method: JStaticMethodID,
}

impl JClassAPI {
    pub fn get(env: &Env<'_>) -> Result<&'static Self> {
        static JCLASS_API: OnceCell<JClassAPI> = OnceCell::new();
        JCLASS_API.get_or_try_init(|| {
            let vm = env.get_java_vm();
            vm.with_env_current_frame(|env| {
                // NB: Self::CLASS_NAME is a binary name with dots, not slashes
                let class = env.find_class(JNIStr::from_cstr(c"java/lang/Class"))?;
                let class = env.new_global_ref(class)?;
                let get_class_loader_method =
                    env.get_method_id(&class, c"getClassLoader", c"()Ljava/lang/ClassLoader;")?;
                let for_name_method = env.get_static_method_id(
                    &class,
                    c"forName",
                    c"(Ljava/lang/String;)Ljava/lang/Class;",
                )?;
                let for_name_with_loader_method = env.get_static_method_id(
                    &class,
                    c"forName",
                    c"(Ljava/lang/String;ZLjava/lang/ClassLoader;)Ljava/lang/Class;",
                )?;
                Ok(Self {
                    class,
                    get_class_loader_method,
                    for_name_method,
                    for_name_with_loader_method,
                })
            })
        })
    }
}

impl JClass<'_> {
    /// Creates a [`JClass`] that wraps the given `raw` [`jclass`]
    ///
    /// # Safety
    ///
    /// `raw` may be a null pointer. If `raw` is not a null pointer, then:
    ///
    /// * `raw` must be a valid raw JNI local reference.
    /// * There must not be any other `JObject` representing the same local reference.
    /// * The lifetime `'local` must not outlive the local reference frame that the local reference
    ///   was created in.
    pub const unsafe fn from_raw(raw: jclass) -> Self {
        Self(JObject::from_raw(raw as jobject))
    }

    /// Returns the raw JNI pointer.
    pub const fn as_raw(&self) -> jclass {
        self.0.as_raw() as jclass
    }

    /// Unwrap to the raw jni type.
    pub const fn into_raw(self) -> jclass {
        self.0.into_raw() as jclass
    }

    /// Returns the class loader for this class.
    ///
    /// This is used to find the class loader that was responsible for loading this class.
    ///
    /// It may return null for bootstrap classes or objects representing primitive types not associated with a class loader.
    ///
    /// # Throws
    ///
    /// `SecurityException` if the class loader cannot be accessed.
    pub fn get_class_loader<'local>(&self, env: &mut Env<'local>) -> Result<JClassLoader<'local>> {
        let api = JClassAPI::get(env)?;

        // Safety: We know that `getClassLoader` is a valid method on `java/lang/Class` that has no
        // arguments and it returns a valid `ClassLoader` instance.
        let loader = unsafe {
            let loader = env
                .call_method_unchecked(self, api.get_class_loader_method, JavaType::Object, &[])?
                .l()?;
            JClassLoader::from_raw(loader.into_raw())
        };
        Ok(loader)
    }

    /// Finds a class by its fully-qualified binary name or array descriptor.
    ///
    /// This is a method binding for `java.lang.Class.forName(String)`
    ///
    /// This method is used to locate a class by its name, which may be either a fully-qualified
    /// binary name (e.g., `java.lang.String`) or an array descriptor (e.g., `[Ljava.lang.String;`).
    ///
    /// Note: that unlike `FindClass` the names use dot (`.`) notation instead of slash (`/`) notation.
    ///
    /// # Throws
    ///
    /// This method may throw a `ClassNotFoundException` if the class cannot be found.
    pub fn for_name<'local, C>(class_name: C, env: &mut Env<'local>) -> Result<JClass<'local>>
    where
        C: AsRef<JNIStr>,
    {
        let api = JClassAPI::get(env)?;

        let class_name = env.new_string(class_name)?;

        // Safety: We know that `forName` is a valid static method on `java/lang/Class` that takes
        // a String and returns a valid `Class` instance.
        let class = unsafe {
            let class = env
                .call_static_method_unchecked(
                    &api.class,
                    api.for_name_method,
                    JavaType::Object,
                    &[JValue::Object(&class_name).as_jni()],
                )?
                .l()?;
            JClass::from_raw(class.into_raw())
        };
        Ok(class)
    }

    /// Finds a class by its fully-qualified binary name or array descriptor.
    ///
    /// This is a method binding for `java.lang.Class.forName(String, boolean, ClassLoader)`
    ///
    /// This method is used to locate a class by its name (via the ClassLoader) which may be either
    /// a fully-qualified binary name (e.g., `java.lang.String`) or an array descriptor (e.g.,
    /// `[Ljava.lang.String;`).
    ///
    /// Note: that unlike `FindClass` the names use dot (`.`) notation instead of slash (`/`) notation.
    ///
    /// If initialized is true, the class will be initialized before it is returned.
    ///
    /// # Throws
    ///
    /// This method may throw a `ClassNotFoundException` if the class cannot be found.
    pub fn for_name_with_loader<'loader_local, 'env_local, C, L>(
        class_name: C,
        initialize: bool,
        loader: L,
        env: &mut Env<'env_local>,
    ) -> Result<JClass<'env_local>>
    where
        C: AsRef<JNIStr>,
        L: AsRef<JClassLoader<'loader_local>>,
    {
        let api = JClassAPI::get(env)?;

        let class_name = env.new_string(class_name)?;

        // Safety: We know that `forName` is a valid static method on `java/lang/Class` that takes
        // a String, initializer boolean and a ClassLoader and returns a valid `Class` instance.
        let class = unsafe {
            let class = env
                .call_static_method_unchecked(
                    &api.class,
                    api.for_name_with_loader_method,
                    JavaType::Object,
                    &[
                        JValue::Object(&class_name).as_jni(),
                        JValue::Bool(initialize).as_jni(),
                        JValue::Object(loader.as_ref()).as_jni(),
                    ],
                )?
                .l()?;
            JClass::from_raw(class.into_raw())
        };
        Ok(class)
    }
}

// SAFETY: JClass is a transparent JObject wrapper with no Drop side effects
unsafe impl Reference for JClass<'_> {
    const CLASS_NAME: &'static JNIStr = JNIStr::from_cstr(c"java.lang.Class");

    type Kind<'env> = JClass<'env>;
    type GlobalKind = JClass<'static>;

    fn as_raw(&self) -> jobject {
        self.0.as_raw()
    }

    fn lookup_class<'caller>(
        env: &Env<'_>,
        _loader_context: LoaderContext,
    ) -> crate::errors::Result<impl Deref<Target = Global<JClass<'static>>> + 'caller> {
        // As a special-case; we ignore loader_context just to be clear that there's no risk of
        // recursion. (`LoaderContext::load_class` depends on the `JClassAPI`)
        let api = JClassAPI::get(env)?;
        Ok(&api.class)
    }

    unsafe fn from_raw<'env>(local_ref: jobject) -> Self::Kind<'env> {
        JClass::from_raw(local_ref)
    }

    unsafe fn from_global_raw(global_ref: jobject) -> Self::GlobalKind {
        JClass::from_raw(global_ref)
    }
}
