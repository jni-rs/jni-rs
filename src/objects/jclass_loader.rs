use std::ops::Deref;

use once_cell::sync::OnceCell;

use crate::{
    env::Env,
    errors::Result,
    objects::{Global, JClass, JMethodID, JObject, JValue, LoaderContext},
    signature::JavaType,
    strings::JNIStr,
    sys::{jclass, jobject},
};

use super::JObjectRef;

/// A `java.lang.ClassLoader` reference
#[repr(transparent)]
#[derive(Debug, Default)]
pub struct JClassLoader<'local>(JObject<'local>);

impl<'local> AsRef<JClassLoader<'local>> for JClassLoader<'local> {
    fn as_ref(&self) -> &JClassLoader<'local> {
        self
    }
}

impl<'local> AsRef<JObject<'local>> for JClassLoader<'local> {
    fn as_ref(&self) -> &JObject<'local> {
        self
    }
}

impl<'local> ::std::ops::Deref for JClassLoader<'local> {
    type Target = JObject<'local>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<'local> From<JClassLoader<'local>> for JObject<'local> {
    fn from(other: JClassLoader<'local>) -> JObject<'local> {
        other.0
    }
}

struct JClassLoaderAPI {
    class: Global<JClass<'static>>,
    load_class_method: JMethodID,
}

impl JClassLoaderAPI {
    fn get(env: &Env<'_>) -> Result<&'static Self> {
        static JCLASS_LOADER_API: OnceCell<JClassLoaderAPI> = OnceCell::new();
        JCLASS_LOADER_API.get_or_try_init(|| {
            let vm = env.get_java_vm();
            vm.with_env_current_frame(|env| {
                // NB: Self::CLASS_NAME is a binary name with dots, not slashes
                let class = env.find_class(c"java/lang/ClassLoader")?;
                let class = env.new_global_ref(&class).unwrap();
                let load_class_method = env.get_method_id(
                    &class,
                    c"loadClass",
                    c"(Ljava/lang/String;)Ljava/lang/Class;",
                )?;
                Ok(Self {
                    class,
                    load_class_method,
                })
            })
        })
    }
}

impl JClassLoader<'_> {
    /// Creates a [`JClass`] that wraps the given `raw` [`jclass`]
    ///
    /// # Safety
    ///
    /// `raw` may be a null pointer. If `raw` is not a null pointer, then:
    ///
    /// * `raw` must be a valid raw JNI local reference.
    /// * There must not be any other `JObject` representing the same local reference.
    /// * The lifetime `'local` must not outlive the local reference frame that the local reference
    ///   was created in.
    pub const unsafe fn from_raw(raw: jclass) -> Self {
        Self(JObject::from_raw(raw as jobject))
    }

    /// Returns the raw JNI pointer.
    pub const fn as_raw(&self) -> jclass {
        self.0.as_raw() as jclass
    }

    /// Unwrap to the raw jni type.
    pub const fn into_raw(self) -> jclass {
        self.0.into_raw() as jclass
    }

    /// Loads a class by name using this class loader.
    ///
    /// This is a Java method binding for `java.lang.ClassLoader.loadClass(String)`.
    ///
    /// # Throws
    ///
    /// `ClassNotFoundException` if the class cannot be found.
    pub fn load_class<'local>(
        &self,
        name: &JNIStr,
        env: &mut Env<'local>,
    ) -> Result<JClass<'local>> {
        let api = JClassLoaderAPI::get(env)?;

        let name = env.new_string(name)?;

        // SAFETY:
        // - we know that `self` is a valid `JClassLoader` reference and `load_class_method` is a valid method ID.
        // - we know that `loadClass` returns a valid `Class` reference.
        let cls_obj = unsafe {
            let cls = env
                .call_method_unchecked(
                    self,
                    api.load_class_method,
                    JavaType::Object,
                    &[JValue::Object(&name).as_jni()],
                )?
                .l()?;
            JClass::from_raw(cls.into_raw())
        };
        Ok(cls_obj)
    }
}

// SAFETY: JClassLoader is a transparent JObject wrapper with no Drop side effects
unsafe impl JObjectRef for JClassLoader<'_> {
    const CLASS_NAME: &'static JNIStr = JNIStr::from_cstr(c"java.lang.ClassLoader");

    type Kind<'env> = JClassLoader<'env>;
    type GlobalKind = JClassLoader<'static>;

    fn as_raw(&self) -> jobject {
        self.0.as_raw()
    }

    fn lookup_class<'env>(
        env: &'env Env<'_>,
        _loader_context: LoaderContext,
    ) -> crate::errors::Result<impl Deref<Target = Global<JClass<'static>>> + 'env> {
        // As a special-case; we ignore loader_context just to be clear that there's no risk of
        // recursion. (`LoaderContext::load_class` depends on the `JClassLoaderAPI`)
        let api = JClassLoaderAPI::get(env)?;
        Ok(&api.class)
    }

    unsafe fn from_raw<'env>(local_ref: jobject) -> Self::Kind<'env> {
        JClassLoader::from_raw(local_ref)
    }

    unsafe fn from_global_raw(global_ref: jobject) -> Self::GlobalKind {
        JClassLoader::from_raw(global_ref)
    }
}
