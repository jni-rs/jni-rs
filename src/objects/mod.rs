mod jobject;
pub use self::jobject::*;

mod jthrowable;
pub use self::jthrowable::*;

mod jclass;
pub use self::jclass::*;

mod jclass_loader;
pub use self::jclass_loader::*;

mod jstring;
pub use self::jstring::*;

mod jcollection;
pub use self::jcollection::*;

mod jset;
pub use self::jset::*;

mod jiterator;
pub use self::jiterator::*;

mod jmap;
pub use self::jmap::*;

mod jlist;
pub use self::jlist::*;

mod jbytebuffer;
pub use self::jbytebuffer::*;

mod jthread;
pub use self::jthread::*;

/// Primitive Array types
mod jobject_array;
pub use self::jobject_array::*;

mod type_array;
pub use self::type_array::*;

/// Primitive Array types
mod jprimitive_array;
pub use self::jprimitive_array::*;

#[doc(hidden)]
#[deprecated(
    since = "0.22.0",
    note = "Please use `jni::JValue*` instead of `jni::objects::JValue*`."
)]
pub use crate::jvalue::*;

#[doc(hidden)]
#[deprecated(
    since = "0.22.0",
    note = "Please use ID types under `jni::ids::*` instead of `jni::objects::*`."
)]
pub use crate::ids::*;

#[doc(hidden)]
#[deprecated(
    since = "0.22.0",
    note = "Please use reference types under `jni::refs::*` instead of `jni::objects::*`."
)]
pub use crate::refs::*;

#[doc(hidden)]
#[deprecated(
    since = "0.22.0",
    note = "Please use array elements types under `jni::elements::*` instead of `jni::objects::*`."
)]
pub use crate::elements::*;
