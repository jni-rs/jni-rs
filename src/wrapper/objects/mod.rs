// wrappers arount jni pointer types that add lifetimes and other functionality.
mod jvalue;
pub use self::jvalue::*;

mod jmethodid;
pub use self::jmethodid::*;

mod jstaticmethodid;
pub use self::jstaticmethodid::*;

mod jfieldid;
pub use self::jfieldid::*;

mod jstaticfieldid;
pub use self::jstaticfieldid::*;

mod jobject;
pub use self::jobject::*;

mod jthrowable;
pub use self::jthrowable::*;

mod jclass;
pub use self::jclass::*;

mod jstring;
pub use self::jstring::*;

mod jmap;
pub use self::jmap::*;

mod jlist;
pub use self::jlist::*;

mod jbytebuffer;
pub use self::jbytebuffer::*;

// For storing a reference to a java object
mod global_ref;
pub use self::global_ref::*;

// For automatic local ref deletion
mod auto_local;
pub use self::auto_local::*;

// For automatic pointer-based byte array deletion
mod auto_byte_array;
pub use self::auto_byte_array::*;

// For automatic pointer-based primitive array deletion
mod auto_primitive_array;
pub use self::auto_primitive_array::*;
