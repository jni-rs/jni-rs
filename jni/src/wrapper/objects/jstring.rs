use objects::JObject;

use sys::{
    jobject,
    jstring,
};

/// Lifetime'd representation of a `jstring`. Just a `JObject` wrapped in a new
/// class.
#[repr(C)]
#[derive(Clone, Copy)]
pub struct JString<'a>(JObject<'a>);

impl<'a> From<jstring> for JString<'a> {
    fn from(other: jstring) -> Self {
        JString(From::from(other as jobject))
    }
}

impl<'a> ::std::ops::Deref for JString<'a> {
    type Target = JObject<'a>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<'a> From<JString<'a>> for JObject<'a> {
    fn from(other: JString) -> JObject {
        other.0
    }
}

impl<'a> From<JObject<'a>> for JString<'a> {
    fn from(other: JObject) -> JString {
        (other.into_inner() as jstring).into()
    }
}
