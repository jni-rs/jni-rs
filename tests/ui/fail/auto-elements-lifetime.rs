#[path = "../../util/mod.rs"]
mod util;
use util::attach_current_thread;

use jni::objects::ReleaseMode;

pub fn main() {
    attach_current_thread(|env0| {
        let smuggle = env0
            .with_local_frame(10, |env1| -> jni::errors::Result<_> {
                let java_array = env1
                    .new_int_array(3)
                    .expect("JNIEnv#new_int_array must create a java array with given size");

                // It should be OK to get AutoElements from a new `env2` frame because
                // it is only constrained by the `env1` lifetime of it's array reference
                let elems1 = env1.with_local_frame(10, |env2| -> jni::errors::Result<_> {
                    let elems2 = unsafe {
                        env2.get_array_elements(java_array, ReleaseMode::CopyBack)
                            .unwrap()
                    };
                    Ok(elems2)
                })?;

                // But the borrow checker should prevent this...
                Ok(elems1)
            })
            .unwrap();

        eprintln!("BUG: AutoElements has out-lived JNI frame of array reference!");
        drop(smuggle);
        Ok(())
    })
    .unwrap();
}
