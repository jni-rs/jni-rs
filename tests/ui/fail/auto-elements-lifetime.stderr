error: lifetime may not live long enough
  --> tests/ui/fail/auto-elements-lifetime.rs:26:17
   |
10 |             .with_local_frame(10, |env1| -> jni::errors::Result<_> {
   |                                    ----     ---------------------- return type of closure is std::result::Result<AutoElements<'2, i32, JPrimitiveArray<'_, i32>>, jni::errors::Error>
   |                                    |
   |                                    has type `&mut JNIEnv<'1>`
...
26 |                 Ok(elems1)
   |                 ^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
